packages/instantsearch.js/src/lib/utils/render-args.ts

```diff
@@ -30,6 +31,10 @@ export function createRenderArgs(
   parent: IndexWidget
 ) {
   const results = parent.getResults()!;
+  // To make the UI optimistic, we will always render using the current state,
+  // but the previous results. This means a change will be visible immediately,
+  // regardless of the status of the network request.
+  results._state = parent.getHelper()!.state;
```

Comment:
instead of patching in render args, we could also patch in `on('results')` or similar, although when I tried that, that was slightly too late (a "result" render would be able to cause a temporary render with result state instead of eager state).

I'm not the biggest fan of doing this patching here, so if another place makes sense, that would be nice.

It can't be the constructor of search results, as that should still be based on the associated _state, as otherwise we're reading from the wrong query (eg previous query had a disjunctive refinement, current state doesn't have that, you don't want to read from first query, but still from second query, only after results are instantiated you patch _state).